Include/Rover/Model.hpp

```diff
@@ -0,0 +1,190 @@
+      const V& v, const B& b) {
+    c = v / b;
+  }
+}
+
+namespace Rover { 
+
+  //! Single-value regression (prediction) model.
+  /*!
+    \tparam A The type of the regression algorithm.
+    \tparam T The type of the trial.
+  */
+  template<typename A, typename T>
+  class Model {
+    public:
+
+      //! The type of the regression algorithm.
+      using Algorithm = A;
+
+      //! The computation type used by the algorithm.
+      using CompType = typename Algorithm::CompType;
+
+      //! The type of the trial.
+      using Trial = T;
+
+      //! The type of a sample.
+      using Sample = typename Trial::Sample;
+
+      //! The type of a sample's arguments.
+      using Arguments = typename Sample::Arguments;
+
+      //! The type of a sample's result.
+      using Result = typename Sample::Result;
+
+      //! Creates a Model.
+      /*!
+        \param trial The trial.
+        \param args The arguments for the algorithm's constructor.
+      */
+      template<typename... AlgoArgFwd>
+      explicit Model(const Trial& trial, AlgoArgFwd&&... args);
+
+      //! Predicts a Result for given arguments.
+      auto operator ()(const Arguments& args) const;
+
+    private:
+      using Reader = ScalarView<CompType>;
+      using ReaderSample = typename Reader::Sample;
+      using ReaderArguments = typename Reader::Sample::Arguments;
+      using ReaderResult = typename Reader::Sample::Result;
+
+      Sample m_basis;
+      Algorithm m_algorithm;
+
+      static Sample compute_basis(const Trial& trial);
+      ReaderSample sample_cast(const Sample& sample) const;
+      ReaderArguments arguments_cast(const Arguments& sample) const;
+      ReaderResult result_cast(const Result& value) const;
+      auto result_cast(ReaderResult value) const;
+  };
+
+  template<typename A, typename T>
+  template<typename... AlgoArgFwd>
+  Model<A, T>::Model(const Trial& trial, AlgoArgFwd&&... args)
+    : m_basis(compute_basis(trial)),
+      m_algorithm(std::forward<AlgoArgFwd>(args)...) {
+    auto reader = ScalarView<CompType>([&](std::size_t i) {
+      return sample_cast(trial[i]);
+    }, trial.size());
+    m_algorithm.learn(std::move(reader));
+  }
+
+  template<typename A, typename T>
+  auto Model<A, T>::operator ()(const Arguments& args)
+      const {
+    auto arguments = arguments_cast(args);
+    auto value = m_algorithm.predict(arguments);
+    auto result = result_cast(value);
+    return result;
+  }
+
+  template<typename A, typename T>
+  typename Model<A, T>::Sample Model<A, T>::compute_basis(const Trial& trial) {
+    auto order = std::vector<std::size_t>(trial.size());
+    auto generator = std::mt19937(std::random_device()());
+    std::iota(order.begin(), order.end(), 0LU);
+    auto basis = *trial.begin();
+    visit_arguments([&](auto& result, auto i) {
+      std::shuffle(order.begin(), order.end(), generator);
+      const auto& first_arguments = trial[order[0]].m_arguments;
```

Comment:
Remove the ```const```. Generally when declaring a variable, adding a ```const``` to ```auto``` doesn't serve much of any purpose.

Reply:
Does it? In this case, `ListTrial::operator[]` and `TrialView::operator[]` only return a const reference, but commonly containers implementing operator[] return non-const references. Binding to a const reference promoted comprehension and streamlines compiler optimization ("use const whenever possible" thing). When I see code like this:
`auto& value = vec[0];`
I expect value to be modified in the code - don't you?
Furthermore, unlike `auto&`, `const auto&` works if `operator[]` returns an rvalue, and it does not seem like an extraordinary use case.

Reply:
Yep good points, it's about maintaining consistency with the rest of the codebase. ```auto&``` will resolve to a ```const&``` in this scenario. The use of ```const``` here has no effect on compiler optimizations.

Reply:
It will but only for ListTrial. More sophisticated Trials will most likely return rvalues because the only reason for using different Trials is to handle cases where all the samples do not reside in one system's RAM all the time. In those cases, binding operator[]'s result to `auto&` won't be possible.
`auto&&` is a viable alternative too.
The same applies to compiler optimizations etc., it does not make a difference if the trial's operator[] returns a constant reference, but does otherwise.
N.B. Current TrialView does not work with non-const-ref trials. It may need to be fixed as well.

Reply:
Thank you for your arguments, if it must work with both lvalues and rvalues then your decision to use ```const auto&``` is correct. Disregard my comment.